### PR TITLE
[0.17] Block v4 mandatory

### PR DESCRIPTION
**Pull request by dongcarl**
By setting the BIPXX consensus fields to `0` we allow them to be enforced from genesis block by default.

This means blocks will be required to be at least version 4 by default.

Also adds tests.

Todo for later: `BIP34_IMPLIES_BIP30_LIMIT` right now means all chains will slow down validation at height `1983702`. We should prod upstream to do something smarter, or allow this to be set as an argument.

### CODE REVIEW

**Comment by dongcarl** [comment]
Made the relevant changes and rebased over https://github.com/ElementsProject/elements/pull/452, having some trouble making tests work. Seems like making `g_signed_blocks` true might fix it, but breaks other tests.

**Comment by instagibbs** [comment]
note that this new method of activating new blocks is an attempt to use the starting block of various rules at block 0. This will need reasonable review.

**Comment by instagibbs** [comment]
For `bitcoin_functional` tests the original regtest activation heights will likely have to be put in the `util.py` function that creates the conf file.

**Comment by instagibbs** [comment]
I believe this will get you to passing tests, but we'll need a test of our own setting them to zero:

```
diff --git a/test/bitcoin_functional/functional/test_framework/util.py b/test/bitcoin_functional/functional/test_framework/util.py
index 9b2129c..145f8ce 100644
--- a/test/bitcoin_functional/functional/test_framework/util.py
+++ b/test/bitcoin_functional/functional/test_framework/util.py
@@ -309,6 +309,9 @@ def initialize_datadir(dirname, n, chain):
         f.write("con_connect_coinbase=0\n")
         f.write("anyonecanspendaremine=0\n")
         f.write("con_blockheightinheader=0\n")
+        f.write("con_bip34height=100000000\n")
+        f.write("con_bip65height=1351\n")
+        f.write("con_bip66height=1251\n")
         os.makedirs(os.path.join(datadir, 'stderr'), exist_ok=True)
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
diff --git a/test/functional/test_framework/util.py b/test/functional/test_framework/util.py
index 7630e32..d250fd1 100644
--- a/test/functional/test_framework/util.py
+++ b/test/functional/test_framework/util.py
@@ -308,6 +308,9 @@ def initialize_datadir(dirname, n, chain):
         f.write("con_blocksubsidy=5000000000\n")
         f.write("con_connect_coinbase=0\n")
         f.write("anyonecanspendaremine=0\n")
+        f.write("con_bip34height=100000000\n")
+        f.write("con_bip65height=1351\n")
+        f.write("con_bip66height=1251\n")
         os.makedirs(os.path.join(datadir, 'stderr'), exist_ok=True)
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
```

**Comment by dongcarl** [comment]
@instagibbs Added tests, lmk if you want something more extensive.

**Comment by instagibbs** [comment]
unmarked WIP

**Comment by instagibbs** [comment]
oh baby that test :1st_place_medal: I gotta learn that mininode stuff too.

Not even a nit per se but why the CLTV test? To make sure that blocks are being made of v4?

**Comment by instagibbs** [comment]
ACK https://github.com/ElementsProject/elements/pull/470/commits/e04f6c7018b6a0d22d1302f32c524620b5c6e83d

**Comment by dongcarl** [comment]
> oh baby that test 1st_place_medal I gotta learn that mininode stuff too.
> 
> Not even a nit per se but why the CLTV test? To make sure that blocks are being made of v4?

Yup exactly! :smile:

**Comment by instagibbs** [comment]
I added a description to the OP

**Comment by dongcarl** [comment]
> I added a description to the OP

Thank you sir!
